### PR TITLE
feat:[PLG-318] Implement Base URL Generation for Microservices

### PR DESCRIPTION
**Pull request by akashpaladin**
# Description
Implement Base URL Generation for Microservices in Shipper Service
[PLG-318] 
Fixes # (issue)
[PLG-318] 
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Chore (no code changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

in-order to test the changes, I've deleted the redhat related data from assetgroup_stats index and made count of assets and violations as 0
image
image
then ran shipper with latest changes, and the data got updated successfully 
image
image

- [ ] Test A
- [ ] Test B

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] My commit message/PR follows the contribution guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# **Other information**:

List any documentation updates that are needed for the Wiki

[PLG-318]: https://paladincloud.atlassian.net/browse/PLG-318?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[PLG-318]: https://paladincloud.atlassian.net/browse/PLG-318?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java [hunk 1]
```diff
@@ -34,6 +34,7 @@ public class HttpUtil {
 
     /** The log. */
     static final Logger LOGGER = LoggerFactory.getLogger(HttpUtil.class);
+    private static final String PACMAN_API_URI = System.getenv("PACMAN_API_URI");
```
[the remark this answers is not in the document]
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java:37]: when I opened all shipper jobs,
PACMAN_API_URI is available for them

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetsCountManager.java [hunk 2]
```diff
@@ -4,15 +4,14 @@
 import com.tmobile.cso.pacman.datashipper.util.AssetGroupUtil;
 import com.tmobile.cso.pacman.datashipper.util.AuthManager;
 import com.tmobile.cso.pacman.datashipper.util.Constants;
+import com.tmobile.cso.pacman.datashipper.util.HttpUtil;
```
Comment (mkinchin-paladin) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetsCountManager.java:7]: Don't think it's needed here.

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/IssueCountManager.java [hunk 3]
```diff
@@ -2,6 +2,7 @@
 import com.microsoft.azure.management.resources.Subscription;
 import com.tmobile.cso.pacman.datashipper.util.AuthManager;
 import com.tmobile.cso.pacman.datashipper.util.Constants;
+import com.tmobile.cso.pacman.datashipper.util.HttpUtil;
```
Comment (mkinchin-paladin) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/IssueCountManager.java:5]: Don't think it's needed here.

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/DatasourceDataFetcher.java [hunk 4]
```diff
@@ -51,6 +51,9 @@ public DatasourceData fetchDatasourceData(String datasource) {
             List<String> assetGroups = getVisibleAssetGroupsFiltered(aliases);
             datasourceData.setAssetGroups(assetGroups);
             datasourceData.setAccountIds(getAccountsByDatasource(datasource));
+            LOGGER.info("Successfully collected asset groups and account IDs");
+            LOGGER.info("Count of asset groups : {} and count of account IDs : {}", assetGroups.size(),
+                    datasourceData.getAccountIds().size());
```
Comment (mkinchin-paladin) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/DatasourceDataFetcher.java:56]: I would suggest to have it as a single message - easier to read logs when they are short.
Like "SUCCESS: Count of asset groups : {} and count of account IDs : {}"

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java [hunk 5]
```diff
@@ -1,384 +1,300 @@
...
         Map<String, Object> typeCountMap = Util.parseJson(typeCountJson);
-        List<Map<String, Object>> typeCounts = new ArrayList<>();
-        try {
-            typeCounts = (List<Map<String, Object>>) ((Map<String, Object>) typeCountMap.get("data")).get("assetcount");
-        } catch (Exception e) {
-            throw e;
-        }
-        return typeCounts;
+
+        return  (List<Map<String, Object>>) ((Map<String, Object>) typeCountMap.get("data")).get("assetcount");
```
Comment (mkinchin-paladin) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java:41]: extra space

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java [hunk 6]
```diff
@@ -1,384 +1,300 @@
...
-            throw e;
-        }
+        String responseJson = HttpUtil.get(api + "/patching?ag=" + ag, AuthManager.getToken());
+        Map<String, Object> vulnMap = Util.parseJson(responseJson);
+
+        @SuppressWarnings("unchecked")
+        Map<String, Map<String, Object>> data = (Map<String, Map<String, Object>>) vulnMap.get("data");
+        Map<String, Object> output = data.get(OUTPUT);
+        if (output != null)
+            patchingInfo.putAll(data.get(OUTPUT));
```
Comment (mkinchin-paladin) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java:61]: add braces

**Review by mkinchin-paladin — CHANGES_REQUESTED** [review]
Minor changes needed

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetsCountManager.java (hunk 2 above)
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetsCountManager.java]: removed

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/IssueCountManager.java (hunk 3 above)
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/IssueCountManager.java]: removed

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/DatasourceDataFetcher.java (hunk 4 above)
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/DatasourceDataFetcher.java]: updated

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java (hunk 5 above)
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java:40]: removed

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java (hunk 6 above)
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/AssetGroupUtil.java]: added

**Review by mkinchin-paladin — CHANGES_REQUESTED** [review]

**Review by mkinchin-paladin — APPROVED** [review]
LGTM!

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java [hunk 7]
```diff
@@ -23,89 +18,92 @@
...
+    private static final String API_URI = System.getenv("PACMAN_API_URI");
 
-    private HttpUtil(){
+    private HttpUtil() {
+        throw new IllegalStateException("Constants is a utility class");
     }
+
     /**
-     * Gets the.
+     * Gets the data from http
```
[the remark this answers is not in the document]
Comment (mkinchin-paladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java:38]: pls disregard this comment and proceed with merge.

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetGroupStatsCollector.java [hunk 8]
```diff
@@ -200,7 +189,8 @@ public void uploadAssetGroupTagCompliance(List<String> assetGroups) throws Excep
         List<Map<String, Object>> docs = new ArrayList<>();
         for (String ag : assetGroups) {
             try {
-                Map<String, Object> doc = AssetGroupUtil.fetchTaggingSummary(COMP_API_URL, ag, getToken());
+                Map<String, Object> doc = AssetGroupUtil.fetchTaggingSummary(
```
[the remark this answers is not in the document]
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetGroupStatsCollector.java]: sure @plyubich

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetGroupStatsCollector.java [hunk 9]
```diff
@@ -234,7 +224,8 @@ public  void uploadAssetGroupRuleCompliance(DatasourceData datasourceData) throw
         datasourceData.getAssetGroups().forEach((ag) -> {
             List<Map<String, Object>> docList = new ArrayList<>();
             try {
-                docList = AssetGroupUtil.fetchPolicyComplianceInfo(COMP_API_URL, ag, domains,getToken());
+                docList = AssetGroupUtil.fetchPolicyComplianceInfo(
```
[the remark this answers is not in the document]
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/entity/AssetGroupStatsCollector.java]: ok

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java]: Nithin told me that, its added from starting

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (akashpaladin) [reply jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java:38]: ok

File: jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java [hunk 10]
```diff
@@ -124,32 +122,44 @@ public boolean isTrusted(X509Certificate[] arg0, String arg1) throws Certificate
...
-            if (httpClient != null) {
-                httpClient.close();
-            }
         }
+
         return json;
     }
+
+    // Method to generate the base URL for the asset service
+    public static String getAssetServiceBaseUrl() {
```
Comment (plyubich) [inline jobs/pacman-data-shipper/src/main/java/com/tmobile/cso/pacman/datashipper/util/HttpUtil.java:152]: these should be part of the service and not http utiils. http utils should be simply getting data

**Pull request merged**